pkcs11/src/new/functions/session_management.rs

```diff
@@ -35,21 +36,67 @@ impl<'a> Session<'a> {
         unsafe { Rv::from(get_pkcs11!(self.client(), C_CloseSession)(self.handle())).into_result() }
     }
 
-    pub fn login(&self, user_type: UserType, pin: &str) -> Result<()> {
+    // Do not fail if the user is already logged in. It happens if another session on the same slot
+    // has already called the log in operation. Record the login call and only log out when there
+    // aren't anymore sessions requiring log in state.
+    pub fn login(&mut self, user_type: UserType, pin: &str) -> Result<()> {
+        let mut logged_sessions = self
+            .client()
+            .logged_sessions
+            .lock()
+            .expect("Logged sessions mutex poisoned!");
+
         //TODO: zeroize after
         let mut pin = CString::new(pin)?.into_bytes();
-        unsafe {
+
+        // Login is called anyway, without checking the count to make sure the pin is
+        // correct. So that a Session can not get logged in with an incorrect pin if there are
+        // already other sessions logged in.
+        match unsafe {
             Rv::from(get_pkcs11!(self.client(), C_Login)(
                 self.handle(),
                 user_type.into(),
                 pin.as_mut_ptr(),
                 pin.len().try_into()?,
             ))
-            .into_result()
+        } {
+            Rv::Ok | Rv::Error(RvError::UserAlreadyLoggedIn) => {
+                if let Some(session_handles) = logged_sessions.get_mut(&self.slot()) {
+                    // It might already been present in if this session already tried to log in.
+                    let _ = session_handles.insert(self.handle());
+                } else {
+                    let mut new_set = HashSet::new();
+                    let _ = new_set.insert(self.handle());
+                    let _ = logged_sessions.insert(self.slot(), new_set);
+                }
```

Comment:
I don't get it, what's the advantage of using this method instead of a plain old `Atomic` counter like in Parsec? Do we actually use that set somewhere? Seems like a bigger effort for both login and logout, and you don't even have to execute `C_Login` all the time

Reply:
Ah, I think with an atomic there's still room for a race condition there. But still, a `Mutex<u32>` or something like that would probably do just fine.

Reply:
You can see the old implementation currently in Parsec, it contains one `Mutex<usize>` on the `Provider` structure [here](https://github.com/parallaxsecond/parsec/blob/a62e8cb9ef5cd6640d0ec8a6996e8e9be96c8d4e/src/providers/pkcs11/mod.rs#L68) and the `Session` structure with the `is_logged_in` bool and the methods [here](https://github.com/parallaxsecond/parsec/blob/a62e8cb9ef5cd6640d0ec8a6996e8e9be96c8d4e/src/providers/pkcs11/utils.rs#L337).
Comparing the two implementations the advantage that I see are:
* because `Session` no longer contains the information if it is logged in or not, the `login`/`login` methods can take an immutable reference (which I forgot in this commit) 😄 
* all the logic structure is in one place instead of two
* I believe (maybe personal opinion) that the new login methods are shorter and easier to understand

Reply:
Fair enough, I don't think there's a big difference performance-wise so might as well leave it like this